File: minimum-spanning-trees/golang/graph.go
package main

import "fmt"

type Vertex struct {
	name                                  string
	predecessor                           *Vertex
	distance, start_time, end_time, color int
}

type Edge struct {
	start, end string
	weight     int
}

type Adjacency struct {
	end    string
	weight int
}

type Graph struct {
	vertices      map[string]*Vertex
	edges         []*Edge
	adjacencyList map[string][]*Adjacency
}

type UnionFind struct {
	names  string
	parent map[string]string
	rank   map[string]int
}

func (graph *Graph) initializeGraph() {
	graph.adjacencyList = make(map[string][]*Adjacency)
	for _, v := range graph.vertices {
		graph.adjacencyList[v.name] = make([]*Adjacency, 0)
	}
	for _, edge := range graph.edges {
		graph.adjacencyList[edge.start] = append(graph.adjacencyList[edge.start], &Adjacency{end: edge.end, weight: edge.weight})
	}
}

func (graph *Graph) printGraph() {
	for _, v := range graph.vertices {
		fmt.Print(v.name, " : ")

		adjacencies := graph.adjacencyList[v.name]
		for i, adjacency := range adjacencies {
			fmt.Printf("%v(%d)", adjacency.end, adjacency.weight)
			if i < len(adjacencies)-1 {
				fmt.Print(" -> ")
			}
		}

		fmt.Print("\n")
	}
}

func (set *UnionFind) initializeUnionSet(names string) {
	set.names = names
	set.parent = make(map[string]string)
	set.rank = make(map[string]int)
	for _, name := range names {
		set.parent[string(name)] = string(name)
		set.rank[string(name)] = 0
	}
}

func (set *UnionFind) findSet(x string) string {
	if x != set.parent[x] {
		set.parent[x] = set.findSet(set.parent[x])
	}

	return set.parent[x]
}

func (set *UnionFind) union(x, y string) {
	rootOfX := set.findSet(x)
	rootOfY := set.findSet(y)

	if x != y {
		if set.rank[rootOfX] < set.rank[rootOfY] {
			set.parent[rootOfY] = rootOfX
		} else if set.rank[rootOfY] < set.rank[rootOfX] {
			set.parent[rootOfX] = rootOfY
		} else {
			set.parent[rootOfY] = rootOfX
			set.rank[rootOfX]++
		}
	}
}

func (set *UnionFind) getGroup(root string) []string {
	var group []string
	for _, name := range set.names {
		if set.findSet(string(name)) == root {
			group = append(group, string(name))
		}
	}
	return group
}

func (set *UnionFind) allGroups() map[string][]string {
	groups := make(map[string][]string)
	for _, name := range set.names {
		root := set.findSet(string(name))
		groups[root] = append(groups[root], string(name))
	}
	return groups
}

func (set *UnionFind) printSet() {
	for root, members := range set.allGroups() {
		fmt.Printf("root=%v: %v\n", root, members)
	}
}

type vertexHeap []*Vertex

func (v vertexHeap) Len() int {
	return len(v)
}

func (v vertexHeap) Less(i, j int) bool {
	return v[i].distance < v[j].distance
}

func (v vertexHeap) Swap(i, j int) {
	v[i], v[j] = v[j], v[i]
}

func (v *vertexHeap) Push(x interface{}) {
	*v = append(*v, x.(*Vertex))
}

func (v *vertexHeap) Pop() interface{} {
	old := *v
	n := len(old)
	x := old[n-1]
	*v = old[0 : n-1]
	return x
}

File: single-source-shortest-paths/python/graph.py
import enum
import sys


class Color(enum.Enum):
    WHITE = "white"
    GRAY = "gray"
    BLACK = "black"


class Vertex:
    def __init__(self, name):
        self.name = name
        self.distance = 0
        self.predecessor = None
        self.start_time = self.finish_time = 0
        self.color = Color.WHITE

    def __lt__(self, other):
        return self.distance < other.distance


class Edge:
    def __init__(self, start, end, weight=1):
        self.start = start
        self.end = end
        self.weight = weight

    def __lt__(self, other):
        return self.weight < other.weight


class Graph:
    def __init__(self, vertices, edges):
        self.vertices = vertices
        self.edges = edges
        self.adjacency_list = {}
        self.__initialize_graph()

    def print_graph(self):
        for k in self.vertices.keys():
            print(k, end=" : ")
            adjacencies = self.adjacency_list[k]
            for i, (adjacency, weight) in enumerate(adjacencies):
                print(
                    f"{adjacency.name}({weight})",
                    end=" -> " if i < len(adjacencies) - 1 else "",
                )

            print("")

    def __initialize_graph(self):
        for k, v in self.vertices.items():
            self.adjacency_list[k] = []

        for edge in self.edges:
            self.adjacency_list[edge.start].append(
                (self.vertices[edge.end], edge.weight)
            )


class UnionFind:
    def __init__(self, names):
        self.names = names
        self.parent = {name: name for name in names}
        self.rank = {name: 0 for name in names}

    def union(self, x, y):
        root_of_x = self.find_set(x)
        root_of_y = self.find_set(y)

        if root_of_x != root_of_y:
            if self.rank[root_of_x] < self.rank[root_of_y]:
                self.parent[root_of_y] = root_of_x
            elif self.rank[root_of_y] < self.rank[root_of_x]:
                self.parent[root_of_x] = root_of_y
            else:
                self.parent[root_of_y] = root_of_x
                self.rank[root_of_x] += 1

    def find_set(self, name):
        if name != self.parent[name]:
            self.parent[name] = self.find_set(self.parent[name])

        return self.parent[name]

    def print_set(self):
        groups = {}
        for x in self.names:
            root = self.find_set(x)
            groups.setdefault(root, []).append(x)
        for root, members in groups.items():
            print(f"root={root}: {members}")


def initialize_single_source(graph, s):
    vertices = graph.vertices
    for v in vertices.values():
        v.distance = sys.maxsize
        v.predecessor = None

    s.distance = 0


def relax(u, v, weight):
    if v.distance > u.distance + weight:
        print(
            f"relax: u = {u.name}, v = {v.name} / before = {v.distance}, after = {u.distance + weight}"
        )

        v.distance = u.distance + weight
        v.predecessor = u
